### ISSUE TITLE
libbpf 1.0: rename btf__get_from_id() and btf__load() APIs

### ISSUE THREAD

**Issue by anakryiko** [libbpf/libbpf#278]
Details: https://github.com/libbpf/libbpf/wiki/Libbpf:-the-road-to-v1.0#btfh-apis

**Comment by qmonnet** [comment]
Looking at bpftool today, I realised that the following works (although `-B` is redundant because of autodetection):

```
bpftool btf dump file /sys/kernel/btf/i2c_smbus -B /sys/kernel/btf/vmlinux
```

... but that the following does not:

```
bpftool btf dump id 6 -B /sys/kernel/btf/vmlinux
```

(where `id 6` points to a module BTF), because we cannot set a `base_btf` in bpftool before retrieving the BTF object with `btf__get_from_id()`, where we do `res = btf_get_from_fd(btf_fd, NULL);` with a fixed `NULL` pointer for `base_btf`.

(This works with the `file` because we can pass the `base_btf` to `btf__parse_split()`.)

Not sure if that was discussed before, but it would be nice to take this chance to pass a `base_btf` pointer to `btf__get_from_id()`'s successor, or to have another way to set it before calling that function.

**Comment by anakryiko** [comment]
> Not sure if that was discussed before, but it would be nice to take this chance to pass a `base_btf` pointer to `btf__get_from_id()`'s successor, or to have another way to set it before calling that function.

Yes, we are lacking btf__get_from_id() variant accepting base_btf. It came up before in the context of iterating all BTF objects in the system when module BTFs are supported. See https://github.com/libbpf/libbpf/issues/314. I think the biggest challenge here is naming ;) It would be great to get help implementing this, thanks!

**Comment by qmonnet** [comment]
Right, I had missed that other issue, apologies. It does express the same concern.

Understood, I'll see if I can find some time to give you a hand on this, but no promise. Maybe next week.

**Comment by qmonnet** [comment]
I worked on this and on #314 (but cannot self-assign the issue, I guess I lack permissions):
- [Patchwork](https://patchwork.kernel.org/project/netdevbpf/list/?series=515513)
- [Lore](https://lore.kernel.org/bpf/20210714141532.28526-1-quentin@isovalent.com/t/#u)

**Comment by qmonnet** [comment]
We merged [v3 of the set](https://git.kernel.org/pub/scm/linux/kernel/git/bpf/bpf-next.git/commit/?id=f309b4ba989d96c192edd0d730d85bc1dd92a64a), thank you Andrii.

The legacy functions have not been marked as deprecated yet, a follow-up for a helper to mark them deprecated for a specified version is in discussion.

**Comment by qmonnet** [comment]
The functions were renamed and the legacy ones are now marked as deprecated as of libbpf v0.6.
Both this issue and #314 can be closed (but I don't have the permission on the repo to close them myself).

**Comment by anakryiko** [comment]
@qmonnet, 0b46b7550560 ("libbpf: Add LIBBPF_DEPRECATED_SINCE macro for scheduling API deprecations") has "Closes: " tag, which will auto-close this PR when that commit is going to be synced to Github. Unfortunately I forgot to add the same for #314, so need to either remember to close it or somehow set up #314 to be closed when #278 is closed.

**Comment by anakryiko** [comment]
I just closed #314 manually to not keep track of that anymore.

**Comment by qmonnet** [comment]
We can manually link PRs to issues to have them closed together, but I don't know of an easy way to close an issue when another _issue_ is closed. So yeah, manual close was probably the best solution.
